core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -536,10 +541,18 @@ class ReplicaManager(val config: KafkaConfig,
            * This can cause a replica to always be out of sync.
            */
           val initialLogEndOffset = localReplica.logEndOffset
+
           val logReadInfo = localReplica.log match {
             case Some(log) =>
               val adjustedFetchSize = if (Topic.isInternal(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
-              log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //Try the read first, this tells us whether we need all of adjustedFetchSize for this partition
+              var fetch = log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //If the partition is marked as throttled, and we are over-quota then exclude it
+              if (quota.isThrottled(TopicAndPartition(topic, partition)) && quota.isQuotaExceeded) //TODO - could also check ISR inclusion??
```

Comment:
Yes - the idea seemed like a good one, but it led to some complexities. The main problem was that the ratio of the RequestSize:QuotaSize correlated with the amount the throttle would be undercut. Also, if you asked for more than the quota in a single request you could never make progress. The upshot was that the behaviour was a little unintuitive, and the simpler mechanism seems to work well on aggregate. 
So after thinking about it for a while I decided to ditch the isExceededBy(bytes) approach. Keep it simple ... and this way it more closely matches the way the follower works more closely. I hope that makes sense.